File: python/0948.py
# Link: https://leetcode.com/problems/bag-of-tokens/description/
# Runtime: 41 ms
# Memory: 16.77 MB


class Solution:
    def bagOfTokensScore(self, tokens: List[int], power: int) -> int:
        current_power, sorted_tokens = power, tokens[:]
        sorted_tokens.sort()

        low, high, score, can_step = 0, len(sorted_tokens) - 1, 0, True
        while (low <= high) and (can_step):
            if sorted_tokens[low] <= current_power:
                current_power -= sorted_tokens[low]
                low += 1
                score += 1
            elif (low < high) and (score > 0):
                current_power += sorted_tokens[high]
                high -= 1
                score -= 1
            else:
                can_step = False

        return score

File: c++/0948.cpp
// Link: https://leetcode.com/problems/bag-of-tokens/description/
// Runtime: 4 ms
// Memory: 13.17 MB

class Solution {
public:
  int bagOfTokensScore(const std::vector<int> &tokens, const int power) {
    int currentPower = power;
    std::vector<int> sortedTokens = tokens;
    std::sort(sortedTokens.begin(), sortedTokens.end());

    int low = 0, high = static_cast<int>(sortedTokens.size() - 1);
    int score = 0;
    bool canStep = true;
    while ((low <= high) && (canStep)) {
      if (sortedTokens[low] <= currentPower) {
        currentPower -= sortedTokens[low];
        score++;
        low++;
      } else if ((low < high) && (score > 0)) {
        currentPower += sortedTokens[high];
        high--;
        score--;
      } else
        canStep = false;
    }

    return score;
  }
};
